lib/ide/parser/src/jsclient.rs

```diff
@@ -52,11 +55,17 @@ impl Client {
     }
 }
 
+
 impl IsParser for Client {
-    fn parse(&mut self, _program: String) -> api::Result<api::Ast> {
-        match parse(_program) {
-            Ok(json_ast) => Err(ParsingError(json_ast)),
-            Err(_)       => Err(api::interop_error(Error::ScalaException())),
-        }
+    fn parse(&mut self, program:String, ids:IDMap) -> api::Result<api::Ast> {
+        let ast = || {
+            let json_ids = serde_json::to_string(&ids)?;
+            let json_ast = parse(program, json_ids)?;
+            let      ast = serde_json::from_str(&json_ast)?;
+
+            Result::Ok(ast)
+        };
+
+        Ok(ast()?)
```

Comment:
What does it do? Why this lambda? Why not just normal lines?

Reply:
Because `?` inside `let ast = .. ` should operate over `Result<T, jsclient::Error>`.

The closure fixes the type inference problem that is cause by the return type `Result<T, api::Error>`.